src/lib/routes/admin-api/project/api-token.ts

```diff
@@ -0,0 +1,214 @@
+            | 'openApiService'
+        >,
+    ) {
+        super(config);
+        this.apiTokenService = apiTokenService;
+        this.accessService = accessService;
+        this.proxyService = proxyService;
+        this.openApiService = openApiService;
+        this.logger = config.getLogger('project-api-token-controller.js');
+
+        this.route({
+            method: 'get',
+            path: PATH,
+            handler: this.getProjectApiTokens,
+            permission: READ_PROJECT_API_TOKEN,
+            middleware: [
+                openApiService.validPath({
+                    tags: ['Projects'],
+                    operationId: 'getProjectApiTokens',
+                    responses: {
+                        200: createResponseSchema('apiTokensSchema'),
+                    },
+                }),
+            ],
+        });
+
+        this.route({
+            method: 'post',
+            path: PATH,
+            handler: this.createProjectApiToken,
+            permission: CREATE_PROJECT_API_TOKEN,
+            middleware: [
+                openApiService.validPath({
+                    tags: ['Projects'],
+                    operationId: 'createProjectApiToken',
+                    requestBody: createRequestSchema('createApiTokenSchema'),
+                    responses: {
+                        201: resourceCreatedResponseSchema('apiTokenSchema'),
+                    },
+                }),
+            ],
+        });
+
+        this.route({
+            method: 'delete',
+            path: PATH_TOKEN,
+            handler: this.deleteProjectApiToken,
+            acceptAnyContentType: true,
+            permission: DELETE_PROJECT_API_TOKEN,
+            middleware: [
+                openApiService.validPath({
+                    tags: ['Projects'],
+                    operationId: 'deleteProjectApiToken',
+                    responses: {
+                        200: emptyResponse,
+                    },
+                }),
+            ],
+        });
+    }
+
+    async getProjectApiTokens(
+        req: IAuthRequest,
+        res: Response<ApiTokensSchema>,
+    ): Promise<void> {
+        const { user } = req;
+        const { projectId } = req.params;
+        const projectTokens = await this.accessibleTokens(user, projectId);
+        this.openApiService.respondWithValidation(
+            200,
+            res,
+            apiTokensSchema.$id,
+            { tokens: serializeDates(projectTokens) },
+        );
+    }
+
+    async createProjectApiToken(
+        req: IAuthRequest,
+        res: Response<ApiTokenSchema>,
+    ): Promise<any> {
+        const createToken = await createApiToken.validateAsync(req.body);
+        const { projectId } = req.params;
+        if (!createToken.project) {
+            createToken.project = projectId;
+        }
+
+        if (
+            createToken.projects.length === 1 &&
+            createToken.projects[0] === projectId
+        ) {
```

Comment:
What happens if these conditions aren't satisfied here? We return 201 Created without actually performing the action in that case?

Reply:
These are there to verify that project level tokens have access to 1 project and matches the path param.  I can respond with 409 when these are not met. wdty?

Reply:
Maybe 400 Bad Request is more appropriate? If these conditions aren't met, then the request isn't valid for this endpoint.

Reply:
ok 400 it is